cmd/whitesourceExecuteScan.go

```diff
@@ -184,65 +314,202 @@ func resolveProjectIdentifiers(config *ScanOptions, utils whitesourceUtils, sys
+		// It will be created by the scan, we retrieve the token again after scanning.
+		if projectToken != "" {
+			log.Entry().Infof("Resolved project token: '%s'..", projectToken)
+			config.ProjectToken = projectToken
+		} else {
+			log.Entry().Infof("Project '%s' not yet present in WhiteSource", fullProjName)
 		}
-		log.Entry().Infof("Resolved project token: '%s'..", projectToken)
-		config.ProjectToken = projectToken
 	}
-	return nil
+
+	return scan.updateProjects(sys)
 }
 
 // executeScan executes different types of scans depending on the scanType parameter.
 // The default is to download the Unified Agent and use it to perform the scan.
-func executeScan(config *ScanOptions, utils whitesourceUtils) error {
+func executeScan(config *ScanOptions, scan *whitesourceScan, utils whitesourceUtils) error {
 	if config.ScanType == "" {
 		config.ScanType = config.BuildTool
 	}
 
 	switch config.ScanType {
-	case "npm":
-		// Execute scan with whitesource yarn plugin
-		if err := executeYarnScan(config, utils); err != nil {
+	case "mta":
+		// Execute scan for maven and all npm modules
+		if err := executeMTAScan(config, scan, utils); err != nil {
 			return err
 		}
-	default:
-		// Download the unified agent jar file if one does not exist
-		if err := downloadAgent(config, utils); err != nil {
+	case "maven":
+		// Execute scan with maven plugin goal
+		if err := executeMavenScan(config, scan, utils); err != nil {
 			return err
 		}
-
-		// Auto generate a config file based on the working directory's contents.
-		// TODO/NOTE: Currently this scans the UA jar file as a dependency since it is downloaded beforehand
-		if err := autoGenerateWhitesourceConfig(config, utils); err != nil {
+	case "npm":
+		// Execute scan with in each npm module using npm.Executor
+		if err := executeNpmScan(config, scan, utils); err != nil {
 			return err
 		}
-
-		// Execute whitesource scan with unified agent jar file
-		if err := executeUAScan(config, utils); err != nil {
+	case "yarn":
+		// Execute scan with whitesource yarn plugin
+		if err := executeYarnScan(config, scan, utils); err != nil {
+			return err
+		}
+	default:
+		// Execute scan with Unified Agent jar file
+		if err := executeUAScan(config, scan, utils); err != nil {
 			return err
 		}
 	}
 	return nil
 }
 
 // executeUAScan executes a scan with the Whitesource Unified Agent.
-func executeUAScan(config *ScanOptions, utils whitesourceUtils) error {
+func executeUAScan(config *ScanOptions, scan *whitesourceScan, utils whitesourceUtils) error {
+	// Download the unified agent jar file if one does not exist
+	if err := downloadAgent(config, utils); err != nil {
+		return err
+	}
+
+	// Auto generate a config file based on the working directory's contents.
+	// TODO/NOTE: Currently this scans the UA jar file as a dependency since it is downloaded beforehand
+	if err := autoGenerateWhitesourceConfig(config, utils); err != nil {
+		return err
+	}
+
```

Comment:
besides downloading the agent we need to download also Java if not available in the runtime container.
Example can be found in the existing groovy step `whitesourceExecuteScan`

This is essential for scanning non Java projects with UA

Reply:
Yes, but I thought the decision was to not use the go-step for UA scans for the time being. The focus of this PR is to bypass the UA for MTA, Maven and NPM, not to fix the UA use-case.